### DISCUSSION TITLE
NTP settings

### DISCUSSION THREAD

**Discussion by dac2020** [arendst/Tasmota#16213] [Q&A] [answered]
I have been trying to set tasmota devices to ntp server but I'm not sure it's working because when they communicate with the broker, they always show the wrong date/time.

>tele/garden/SENSOR {"Time":"1970-01-02T04:11:38"

DNS and GW are being handed to devices using DHCP.

The broker does have ntp set but I've not found a way to set the time on tasmota devices looking here.
https://tasmota.github.io/docs/Commands/#management

Now, I also read in another post that someone seemed to imply that it doesn't matter what time the tasmota devices have because the broker handles the date/time of the messages. Typically, if I have something in different states for example, I'll use UTC time then convert that later so I always know what the actual event time of something remote was. 

I think I'm missing something because searching the net and looking at the commands, setting NTP and time/date on devices seems to be a possibility.

**Comment by sfromis** [comment, marked as answer]
Maybe tell your specific steps and results, basically stating no more than it did not work is not any info. Beware that you cannot rely on DHCP suggested NTP servers, you have to specifically config in Tasmota. The MQTT broker is in no way involved in NTP, that would be a separate service, maybe on the same box, maybe allowing this to come from the public Internet.

Tasmota time matters very much in some circumstances, but not used in other situations. Specifically about MQTT broker and date/time, other subscribers picking up messages are able to keep track of when they got a message, without a timestamp included in the message itself being what they use. Still, for precise timing, without the transmission delay, I'd normally prefer the time stamp in the message itself. 

And without real time source, Tasmota may not work as well, even outside of the matter of having local time.

**Reply by barbudor** [reply]
Tasmota tries continuously to contact a NTP server
It used to be a blocking process stucking Tasmota
This have been improved but still Tasmota tries every few ten of seconds to get a NTP sync.
As there is no way to disable it, the best is to make it working

As usual, Weblog 3 or Weblog 4 will give more details on what is failing. If you share such logs we could help you on understanding them

**Comment by dac2020** [comment]
Sorry, I didn't include the steps since I thought you'd know I took them as I mentioned it.

$ mosquitto_pub -h 127.0.0.1 -t cmnd/garden/Time -m 0
$ mosquitto_pub -h 192.168.1.134 -t cmnd/garden/NtpServer1 -m 0.us.pool.ntp.org
$ mosquitto_pub -h 192.168.1.134 -t cmnd/garden/NtpServer2 -m 1.us.pool.ntp.org
$ mosquitto_pub -h 127.0.0.1 -t cmnd/garden/Timezone -m -7:00

I always set ntp on devices and servers when ever possible as it makes things a lot easier.

I did set logging to level 3 but it doesn't help since there is no actual error or problem. The device are showing 1070 as a date so isn't that a config issue, something I need to update on the device/s?

**Reply by barbudor** [reply]
You are still mistaking ip of Tasmota and ip of the NTP broker
Why bothering doing that through MQTT instead of console?
It works but that's easier in the console as you have direct feedback

**Comment by dac2020** [comment]
Because I thought once you have the device communicating with the broker, you can control everything from that point on through the broker. Once installed, I would not always have access to the devices but thought I would always be able to get status and send commands through the broker.

**Reply by sfromis** [reply]
One key value of playing around in the console instead of via MQTT is that it is much easier to see what happens, not all messages are published via MQTT. You can get results, error messages and debug info. And it is also the docs we ask for in case of trouble, much more useful than commands without results.

Commands via MQTT is excellent once you have made it work, not so much when testing out to try to making things work. In that phase, do stick to the console (or serial).

**Comment by dac2020** [comment]
I understand that it's easier but I was trying to learn how to control the device fully via the broker once it's set up.
I know how it works from the console since I did those things when I got my first device up and running :).

**Reply by sfromis** [reply]
It has nothing to do with you knowing how the console works. If you want to ask for help, do make sure to get the docs, it is a waste of time to do testing setup via MQTT and not care about docs. This is not about controlling a device, it is about debugging what's screwed up.

**Comment by dac2020** [comment]
There is nothing screwed up that I can tell and I've said many times that I do search the net, try things before asking for help.
I can already control and set the device form http and curl and console, I'm wanting to learn how to do it from the broker now.
Why do you think something is screwed up? It's working as it should be, I just don't know the commands to use when using the broker.

**Reply by barbudor** [reply]
The NTP setttings is screwed up as you don't get it to work and get a proper time, which as we explain is almost mandatory for Tasmota
Until you answer to our question as we asked (WebLog 3 or 4 and provide the logs), then this discussion is going nowhere
Don't try 2 different things at the same time (debugging NTP and learning a new mode of control)

**Comment by dac2020** [comment]
Just because I make some mistakes here and there doesn't mean I'm not listening to your input but please keep in mind, I'm doing this while I'm also working when I get a few minutes here, there :).

Also, you told me this the other day which is why I'm wanting to learn about MQTT.
https://github.com/arendst/Tasmota/discussions/16154#discussioncomment-3328130

Yes, I enabled weblog 3 but didn't share anything yet since I think I know what's wrong and was trying to figure that out.

stat/garden/LOGGING 19:45:06.553 DNS: Unable to resolve '0.us.pool.ntp.org'
stat/garden/LOGGING 19:46:06.710 DNS: Unable to resolve '1.us.pool.ntp.org'
stat/garden/LOGGING 19:47:06.844 DNS: Unable to resolve '0.nl.pool.ntp.org'

I suppose I had reset the device at some point. The device still has the initial ntp server that seems to be set in tasmota but DNS is not resolving. I think once I solve that, it should be good.

Other than DNS, I don't see any other issues in the log at least. I can set that from the console but was trying to see how I could do it from the broker.

'''
tele/garden/LWT Online
cmnd/garden/POWER (null)
tele/garden/INFO1 {"Info1":{"Module":"ESP32-DevKit","Version":"12.0.2(tasmota)","FallbackTopic":"cmnd/DVES_6D54E0_fb/","GroupTopic":"cmnd/tasmotas/"}}
tele/garden/INFO2 {"Info2":{"WebServerMode":"Admin","Hostname":"garden-5344","IPAddress":"192.168.254.201"}}
tele/garden/INFO3 {"Info3":{"RestartReason":"Software reset CPU","BootCount":13}}
stat/garden/RESULT {"POWER1":"OFF"}
stat/garden/POWER1 OFF
stat/garden/RESULT {"POWER2":"OFF"}
stat/garden/POWER2 OFF
stat/garden/RESULT {"POWER3":"OFF"}
stat/garden/POWER3 OFF
stat/garden/RESULT {"POWER4":"OFF"}
stat/garden/POWER4 OFF
stat/garden/LOGGING 00:00:00.003 HDW: ESP32-D0WD
stat/garden/LOGGING 00:00:00.003 HDW: FoundPSRAM=0 CanUsePSRAM=0
stat/garden/LOGGING 00:00:00.079 UFS: FlashFS mounted with 304 kB free
stat/garden/LOGGING 00:00:00.154 CFG: Loaded from File, Count 156
stat/garden/LOGGING 00:00:00.169 QPC: Count 1
stat/garden/LOGGING 00:00:00.171 CFG: CR 458/699, Busy 0
stat/garden/LOGGING 00:00:00.184 ROT: Mode 1
stat/garden/LOGGING 00:00:00.357 CFG: No '*.autoconf' file found
stat/garden/LOGGING 00:00:00.363 BRY: Berry initialized, RAM used=3731 bytes
stat/garden/LOGGING 00:00:00.381 BRY: No 'preinit.be'
stat/garden/LOGGING 00:00:00.385 SRC: Restart
stat/garden/LOGGING 00:00:00.434 Project tasmota - 4chRelay-ch-01 Version 12.0.2(tasmota)-2_0_3(2022-06-20T12:41:24)
stat/garden/LOGGING 00:00:00.437 ETH: No ETH MDC and/or ETH MDIO GPIO defined
stat/garden/LOGGING 00:00:00.449 SHT: Sensor did not ACK command
stat/garden/LOGGING 00:00:00.604 BRY: No 'autoexec.be'
stat/garden/LOGGING 00:00:00.919 WIF: Attempting connection...
stat/garden/LOGGING 00:00:01.251 WIF: Connecting to AP1 gst2 Channel 11 BSSId A4:2B:B0:FA:3C:AB in mode 11n as garden-5344...
stat/garden/LOGGING 00:00:01.331 WIF: Attempting connection...
stat/garden/LOGGING 00:00:03.413 WIF: Connected
stat/garden/LOGGING 00:00:03.679 HTP: Web server active on garden-5344 with IP address 192.168.254.201
stat/garden/LOGGING 00:00:06.884 DNS: Unable to resolve '0.us.pool.ntp.org'
stat/garden/LOGGING 00:00:10.032 DNS: Unable to resolve '1.us.pool.ntp.org'
stat/garden/LOGGING 00:00:13.160 DNS: Unable to resolve '0.nl.pool.ntp.org'
stat/garden/LOGGING 00:00:13.237 QPC: Reset
stat/garden/LOGGING 00:00:16.284 DNS: Unable to resolve '2.pool.ntp.org'
stat/garden/LOGGING 00:00:16.359 MQT: Attempting connection...
stat/garden/LOGGING 00:00:16.385 MQT: Connected
stat/garden/LOGGING 00:00:16.391 MQT: tele/garden/LWT = Online (retained)
stat/garden/LOGGING 00:00:16.394 MQT: cmnd/garden/POWER =
stat/garden/LOGGING 00:00:16.396 MQT: Subscribe to cmnd/garden/#
stat/garden/LOGGING 00:00:16.399 MQT: Subscribe to cmnd/tasmotas/#
stat/garden/LOGGING 00:00:16.401 MQT: Subscribe to cmnd/DVES_6D54E0_fb/#
stat/garden/LOGGING 00:00:16.406 MQT: tele/garden/INFO1 = {"Info1":{"Module":"ESP32-DevKit","Version":"12.0.2(tasmota)","FallbackTopic":"cmnd/DVES_6D54E0_fb/","GroupTopic":"cmnd/tasmotas/"}}
stat/garden/LOGGING 00:00:16.440 MQT: tele/garden/INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"garden-5344","IPAddress":"192.168.254.201"}}
stat/garden/LOGGING 00:00:16.468 MQT: tele/garden/INFO3 = {"Info3":{"RestartReason":"Software reset CPU","BootCount":13}}
stat/garden/LOGGING 00:00:16.485 MQT: stat/garden/RESULT = {"POWER1":"OFF"}
stat/garden/LOGGING 00:00:16.488 MQT: stat/garden/POWER1 = OFF
stat/garden/LOGGING 00:00:16.495 MQT: stat/garden/RESULT = {"POWER2":"OFF"}
stat/garden/LOGGING 00:00:16.505 MQT: stat/garden/POWER2 = OFF
stat/garden/LOGGING 00:00:16.514 MQT: stat/garden/RESULT = {"POWER3":"OFF"}
stat/garden/LOGGING 00:00:16.517 MQT: stat/garden/POWER3 = OFF
stat/garden/LOGGING 00:00:16.520 MQT: stat/garden/RESULT = {"POWER4":"OFF"}
stat/garden/LOGGING 00:00:16.523 MQT: stat/garden/POWER4 = OFF
stat/garden/LOGGING 00:00:16.534 TFS: File 'autoexec.bat' not found
stat/garden/LOGGING 00:00:19.755 DNS: Unable to resolve '0.us.pool.ntp.org'
stat/garden/LOGGING 00:00:19.837 APP: Boot Count 13
stat/garden/LOGGING 00:00:22.893 DNS: Unable to resolve '1.us.pool.ntp.org'
stat/garden/LOGGING 00:00:23.249 CFG: Saved, Count 157, Bytes 4096
stat/garden/LOGGING 00:00:26.321 DNS: Unable to resolve '0.nl.pool.ntp.org'
stat/garden/LOGGING 00:00:30.188 DNS: Unable to resolve '2.pool.ntp.org'
tele/garden/STATE {"Time":"1970-01-01T00:00:30","Uptime":"0T00:00:13","UptimeSec":13,"Heap":153,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":2263,"MqttCount":1,"Berry":{"HeapUsed":3,"Objects":40},"POWER1":"OFF","POWER2":"OFF","POWER3":"OFF","POWER4":"OFF","Wifi":{"AP":1,"SSId":"gst2","BSSId":"A4:2B:B0:FA:3C:AB","Channel":11,"Mode":"11n","RSSI":96,"Signal":-52,"LinkCount":1,"Downtime":"0T00:00:04"}}
tele/garden/SENSOR {"Time":"1970-01-01T00:00:30","ESP32":{"Temperature":88.0},"TempUnit":"F"}
stat/garden/LOGGING 00:00:30.281 MQT: tele/garden/STATE = {"Time":"1970-01-01T00:00:30","Uptime":"0T00:00:13","UptimeSec":13,"Heap":153,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":2263,"MqttCount":1,"Berry":{"HeapUsed":3,"Objects":40},"POWER1":"OFF","POWER2":"OFF","POWER3":"OFF","POWER4":"OFF","Wifi":{"AP":1,"SSId":"gst2","BSSId":"A4:2B:B0:FA:3C:AB","Channel":11,"Mode":"11n","RSSI":96,"Signal":-52,"LinkCount":1,"Downtime":"0T00:00:04"}}
stat/garden/LOGGING 00:00:30.332 MQT: tele/garden/SENSOR = {"Time":"1970-01-01T00:00:30","ESP32":{"Temperature":88.0},"TempUnit":"F"}
stat/garden/LOGGING 00:00:33.406 DNS: Unable to resolve '0.us.pool.ntp.org'
stat/garden/LOGGING 00:00:36.558 DNS: Unable to resolve '1.us.pool.ntp.org'
stat/garden/LOGGING 00:00:39.686 DNS: Unable to resolve '0.nl.pool.ntp.org'
stat/garden/LOGGING 00:00:42.835 DNS: Unable to resolve '0.pool.ntp.org'
'''

**Reply by dac2020** [reply]
While it's ok to set these things from the console, my thinking is what if the device was installed somewhere else and I could not reach the console. I might be able to at least set some things up remotely using the broker.

**Reply by dac2020** [reply]
So I went ahead and set DNS via console
EthDnsServer1 192.168.1.29
EthDnsServer2 192.168.1.30
But still seeing unable to resolve. 
Doesn't seem to be a network issue since it can reach the broker. Something simple I'm not noticing.

**Reply by sfromis** [reply]
Now you finally posted the secret you've kept for so long. `DNS: Unable to resolve` 

That has absolutely nothing to do with NTP settings, but clearly a more general issue with networking. Tasmota can do nothing but send the name resolution requests to the DNS server. If that fails, Tasmota can do no more. The ability to talk to the MQTT broker is in no way enough to conclude that you have no network issue.

Do not waste your time with `EthDnsServer` when (it appears) you have no Ethernet connection yet. You are trying too many steps at once, and likely confusing yourself. I'd suggest starting from scratch with simple steps, and only build on top of that when you have each step working without issues.

Operating via MQTT is fine, but only after you've made a working setup. It is a bad waste of time to "force" initial debugging via MQTT.

**Comment by dac2020** [comment]
Device 'Information' shows correct gateway, dns server IPs, etc. All looks fine and ntp servers are now resolving yet the year remains 1970.
The timezone is set and correct. Still looking on the net because I'm not seeing any obvious errors in the log.

**Comment by dac2020** [comment]
I think we are out of sync.
Yes, I have a DNS problem and that's what I'm trying to solve which should fix the NTP issue.
The wireless is separated from the lan but it should have dns access to will look at this tomorrow.

**Comment by dac2020** [comment]
>That has absolutely nothing to do with NTP settings

It does, DNS is not resolving.

>Do not waste your time with EthDnsServer when (it appears) you have no Ethernet connection yet.

I already showed I am able to reach the broker server which is on 192.168.1.134.
The devices are on a wifi network, 192.168.254.x. 
Nothing wrong with the networking, it's a DNS issue.

>I'd suggest starting from scratch with simple steps, and only build on top of that when you have each step working 
>without issues.

I agree and it's what I've been doing based on a lot of your input.

I can see the correct gw in the 'Information' but I cannot ping it from the console. 
And of course, I cannot ping the DNS servers either which are on 192.168.1.x.

I can ping 4.2.2.2 over the Internet so would seem it's an internal wifi to lan issue.
However, even setting the device DNS server to public 8.8.8.8 is still not resolving.

**Reply by sfromis** [reply]
DNS failure **is** a networking issue. Again, you being able to reach your broker from Tasmota does **by far** not mean that you can rule out networking issues. Your networking is screwed up if you cannot resolve DNS. If you cannot even make 8.8.8.8 work for DNS, you have some networking issue around routing, or whatever.

The point of not wasting time using `EthDnsServer` was simply that this command is only relevant when you have wired Ethernet connection running instead of Wifi. And the messages in the log you posted indicated that this was not the case.

Your networking issues will have to be resolved outside of Tasmota.

**Comment by dac2020** [comment]
I have not ruled anything out yet, I'm still looking to find the problem.
Yes, this is a wireless network and yes, this has to be resolved to move forward :).
Your point that EthDnsServer would not work helps to eliminate that at least.

**Comment by dac2020** [comment]
Just an FYI. I have not abandoned this. The network problem has lead down a rabbit hole of other problems. 
I'll make a small wireless router the mqtt so that it's on the same network and won't have the DNS problem anymore.
However, looking into this caused me to want to make some significant changes so it's taking a little time.

I think this question is solved as a network problem, more specifically, devices are not able to reach dns servers let alone the gateway their DHCP is on.

**Comment by dac2020** [comment]
Sorry, forgot to update this. 
DNS requests were not getting from one network segment to another and that's fixed now.
Thanks for your help.
